src/main/java/bio/terra/validation/StudyRequestValidator.java

```diff
@@ -0,0 +1,203 @@
+        this.pdao = pdao;
+        this.studyDao = studyDao;
+    }
+
+    @Override
+    public boolean supports(Class<?> clazz) {
+        return StudyRequestModel.class.equals(clazz);
+    }
+
+    private static class SchemaValidationContext {
+
+        private HashMap<String, HashSet<String>> tableColumnMap;
+        private HashSet<String> relationshipNameSet;
+
+        SchemaValidationContext() {
+            tableColumnMap = new HashMap<>();
+            relationshipNameSet = new HashSet<>();
+        }
+
+        void addTable(String tableName, List<String> columnNames) {
+            tableColumnMap.put(tableName, new HashSet<>(columnNames));
+        }
+
+        void addRelationship(String relationshipName) {
+            relationshipNameSet.add(relationshipName);
+        }
+
+        boolean isInvalidTable(String tableName) {
+            return !tableColumnMap.containsKey(tableName);
+        }
+
+        boolean isInvalidTableColumn(String tableName, String columnName) {
+            return isInvalidTable(tableName) || !tableColumnMap.get(tableName).contains(columnName);
+        }
+
+        boolean isInvalidRelationship(String relationshipName) {
+            return !relationshipNameSet.contains(relationshipName);
+        }
+    }
+
+    private void validateStudyName(String studyName, Errors errors) {
+        if (studyName == null) {
+            errors.rejectValue("name", "StudyNameMissing");
+        } else if (!Utils.isValidName(studyName)) {
+            errors.rejectValue("name", "StudyNameInvalid");
+        } else if (pdao.studyExists(studyName) /* || TODO: use studyDao to check existence */) {
+            errors.rejectValue("name", "StudyNameInUse");
+        }
+    }
+
+    private void validateTable(TableModel table, Errors errors, SchemaValidationContext context) {
+        String tableName = table.getName();
+        List<ColumnModel> columns = table.getColumns();
+
+        if (tableName != null && columns != null) {
+            List<String> columnNames = columns.stream().map(ColumnModel::getName).collect(toList());
+            if (Utils.hasDuplicates(columnNames)) {
+                errors.rejectValue("schema", "DuplicateColumnNames");
+            }
+            context.addTable(tableName, columnNames);
+        }
+    }
+
+    private void validateRelationshipTerm(RelationshipTermModel term, Errors errors, SchemaValidationContext context) {
+        String table = term.getTable();
+        String column = term.getColumn();
+        if (table != null && column != null) {
+            if (context.isInvalidTableColumn(table, column)) {
+                errors.rejectValue("schema", "InvalidRelationshipTermTableColumn");
+            }
+        }
+    }
+
+    private void validateRelationship(RelationshipModel relationship, Errors errors, SchemaValidationContext context) {
+        RelationshipTermModel fromTerm = relationship.getFrom();
+        if (fromTerm != null) {
```

Comment:
These will get caught by the annotated constraints in the RelationshipTermModel based on what we said was required in the swagger yaml. This validator still runs because Spring validation does not shortcut out early after it finds an issue. I'll add a comment about this at the top of the file.